### DISCUSSION TITLE
Enable Link-Time Optimization (LTO) for CLI

### DISCUSSION THREAD

**Discussion by zamazan4ik** [DOLPHIN-SOFTWARE-TECH/cargo-version-upgrade#3] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a little bit. If you want to read more about LTO and its possible modes, I recommend starting from [this](https://doc.rust-lang.org/rustc/codegen-options/index.html#lto) Rustc documentation.

I think you can enable LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where in addition to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most optimized version of the application. However, if we enable it on the Cargo profile level for the Release profile, users, who install the application with `cargo install` will get the LTO-optimized version of the game "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48). You also could be interested in other optimization options like `codegen-units = 1`, etc.

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

I have made quick tests (Fedora 41, Rust 1.82, the latest version of the project) - here are the results (`ls -la`):
* Release: 1499776 byte
* Release + `codegen-units = 1`: 1300440 byte
* Release + LTO: 1233520 byte
* Release + `codegen-units = 1` + LTO: 1172080 byte

Thank you. It would be nice if you would use LTO for your future Rust developments :)

**Comment by narkreeta** [comment]
@zamazan4ik that's actually good point and thanks for share this information. I will add another args in cli such as `--lto` , so if user want to enable it they can use that args during version upgradeing or just simply can run it.
